tracing/src/main/java/com/palantir/tracing/Tracer.java

```diff
@@ -230,6 +278,31 @@ static DetachedSpan detachInternal(
         return Optional.empty();
     }
 
+    private static Optional<String> getRequestId(@Nullable Trace maybeCurrentTrace, SpanType newSpanType) {
+        if (maybeCurrentTrace != null) {
+            return maybeCurrentTrace.getRequestId();
+        }
+        if (newSpanType == SpanType.SERVER_INCOMING) {
+            return Optional.of(Tracers.randomId());
+        }
+        return Optional.empty();
+    }
+
+    static Optional<String> getRequestId(DetachedSpan detachedSpan) {
+        if (detachedSpan instanceof SampledDetachedSpan) {
+            return ((SampledDetachedSpan) detachedSpan).requestId;
+        }
+        if (detachedSpan instanceof UnsampledDetachedSpan) {
+            return ((UnsampledDetachedSpan) detachedSpan).requestId;
+        }
+        throw new SafeIllegalStateException("Unknown span type", SafeArg.of("detachedSpan", detachedSpan));
+    }
+
+    /** Returns the unique request identifier for this thread's trace. */
+    public static Optional<String> getRequestId() {
+        return checkNotNull(currentTrace.get(), "There is no trace").getRequestId();
```

Comment:
I don't want requestId to be part of the public API here, perhaps this can be hidden.